make_prg/subcommands/update.py

```diff
@@ -0,0 +1,187 @@
+import multiprocessing
+import os
+from pathlib import Path
+from loguru import logger
+from make_prg.utils import io_utils, gfa
+from make_prg.update.denovo_variants import DenovoVariantsDB
+from make_prg.prg_builder import PrgBuilderZipDatabase, LeafNotFoundException
+from make_prg.utils.msa_aligner import MAFFT, MSAAligner
+from dataclasses import dataclass
+
+
+def register_parser(subparsers):
+    subparser_update_prg = subparsers.add_parser(
+        "update",
+        usage="make_prg update",
+        help="Update PRGs given new sequences.",
+    )
+
+    def check_if_is_update_file(argument: str):
+        argument = Path(argument)
+        is_zip_file = argument.suffix == ".zip"
+        is_update_DS = argument.suffix == ".update_DS"
+        if not is_zip_file and not is_update_DS:
+            subparser_update_prg.error(f"{argument} is not a update_DS nor a zip file.")
+        return argument
+    subparser_update_prg.add_argument(
+        "-u",
+        "--update_DS",
+        action="store",
+        type=lambda argument: check_if_is_update_file(argument),
+        required=True,
+        help=(
+            "Filepath to the update data structures (a *.update_DS.zip file created "
+            "by make_prg from_msa)"
+        ),
+
+    )
+    subparser_update_prg.add_argument(
+        "-d",
+        "--denovo_paths",
+        action="store",
+        type=str,
+        required=True,
+        help=(
+            "Filepath containing denovo sequences. Should point to a denovo_paths.txt file"
+        ),
+    )
+    subparser_update_prg.add_argument(
+        "-o",
+        "--output_prefix",
+        action="store",
+        type=str,
+        required=True,
+        help="Prefix for the output files",
+    )
+    subparser_update_prg.add_argument(
+        "-t",
+        "--threads",
+        action="store",
+        type=int,
+        default=1,
+        help="Number of threads (default 1)",
+    )
+    subparser_update_prg.add_argument(
+        "--mafft",
+        help="Path to MAFFT executable. By default, it is assumed to be on $PATH",
+        default="mafft",
+    )
+    subparser_update_prg.add_argument(
+        "--output_graphs",
+        dest="output_graphs",
+        action="store_true",
+        default=False,
+        help="Outputs the recursive tree graphical representation (for development use only)",
+    )
+    subparser_update_prg.set_defaults(func=run)
+
+    return subparser_update_prg
+
+
+@dataclass
+class UpdateSharedData:
+    denovo_variants_db: DenovoVariantsDB
+    aligner: MSAAligner
+
+
+def update(locus_name: str):
+    global options
+    prg_builder_zip_db = PrgBuilderZipDatabase(options.update_DS)
+    prg_builder_zip_db.load()
+    prg_builder_for_locus = prg_builder_zip_db.get_PrgBuilder(locus_name)
+
+    global update_shared_data
```

Comment:
There is pretty excessive use of global in this PR. Is this fixed in a future PR?

Reply:
I don't have any intention to fix this in future PRs, because I don't actually know how to fix this without impacting performance, and I prefer to use a few global variables to improve performance. I will make my point clearer.

As far as I know, `python` does not really implement multithreading due to the global interpreter lock - even if you create multiple threads, only one will run at a time. There are ways around this I guess, but will cause more trouble than what it solves, like using non-cpython implementations or trying to disable the GIL, but this can break packages and modules, so I don't think it is worth trying (see https://wiki.python.org/moin/GlobalInterpreterLock). Multithreading would be great, because we would not need global variables, and we could control access to them, but is not an option.

Our only option is thus multiprocessing. We could indeed avoid the global variables by passing them to the target function of `pool.map()` but this incurs into some overhead, as these variable will have to be pickled by the parent process and unpickled by the child process, thus I chose to use this only for light-data variables, which is not the case of `update_shared_data`, which will hold the whole information of all denovo updates that need to be done to the PanRG. However, this data will only be read by the children processes, and never written by them, so if we declare this variable as global, children processes can access this data from the parent process very efficiently due to [linux using Copy-on-write optimisation when forking processes](https://en.wikipedia.org/wiki/Copy-on-write).

In conclusion, I did not find a way to use threads themselves and control access to the variable, so had to forcibly use multiprocessing. And the most efficient way that I found to share heavy data that is created in the parent process with their children is to use a global variable and rely on the Copy-on-write optimisation. I do agree that using global variables is not a good engineering approach and should be always avoided, but I don't see any better option right now, but would be really happy to hear if you know a better solution!

Reply:
Ahhh I see; didn't realise it was a multithreading thing. All good.